src/catalog/catalog_entry/duck_table_entry.cpp

```diff
@@ -809,6 +781,92 @@ unique_ptr<CatalogEntry> DuckTableEntry::DropForeignKeyConstraint(ClientContext
 	return make_uniq<DuckTableEntry>(catalog, schema, *bound_create_info, storage);
 }
 
+void RollbackAddIndex(CatalogEntry &rollback_entry, CatalogEntry &entry) {
+	auto &rollback_table = rollback_entry.Cast<DuckTableEntry>();
+	auto &table = entry.Cast<DuckTableEntry>();
+	auto &info = table.GetStorage().GetDataTableInfo();
+	auto &indexes = info->GetIndexes();
+
+	// Find all index-based constraints that exist in rollback_table, but not in table.
+	// Then, remove them.
+
+	unordered_set<string> names;
+	for (const auto &constraint : table.GetConstraints()) {
+		if (constraint->type != ConstraintType::UNIQUE) {
+			continue;
+		}
+		const auto &unique = constraint->Cast<UniqueConstraint>();
+		if (unique.is_primary_key) {
+			names.insert(unique.info.name);
+		}
+	}
+
+	for (const auto &constraint : rollback_table.GetConstraints()) {
+		if (constraint->type != ConstraintType::UNIQUE) {
+			continue;
+		}
+		const auto &unique = constraint->Cast<UniqueConstraint>();
+		if (!unique.IsPrimaryKey()) {
+			continue;
+		}
+		if (names.find(constraint->info.name) == names.end()) {
+			indexes.RemoveIndex(constraint->info.name);
+		}
+	}
```

Comment:
Should we add a `FIXME` here that currently this only works for `PRIMARY KEY`, as that's the only index-based constraint that we can currently alter?